recipe-server/normandy/base/genome.py

```diff
@@ -0,0 +1,138 @@
+import json
+import math
+import random
+import string
+
+from hashlib import sha256
+
+
+class Genome:
+    """A seedable source that provides arbitrarily sized chunks of randomness."""
+    def __init__(self, chromosome, entropy):
+        self.chromosome = chromosome
+        self.entropy = entropy
+        self.log = []
+
+    @staticmethod
+    def generate(seed=random.random()):
+        """Asynchronously returns a Genome based on the given seed.
+
+        seed -- Any JSON serializable value. Will be used to seed the
+                genome. If not provided, will be randomly generated.
+        """
+        hash = sha256(json.dumps(seed).encode('utf-8'))
+        chromosome = int(hash.hexdigest(), base=16)
+        return Genome(chromosome, 256)
+
+    def take(self, mod, reason):
+        """Return an integer in the range [0, mod), and deduct the correct
+        amount of entropy from the pool. Vulnerable to modulus-bias when
+        `mod` is not a power of two.
+        This is the fundamental way to consume a Genome. All the other
+        methods below that use entropy defer to this method.
+        """
+        needed = math.log2(mod)
+        if self.entropy < needed:
+            raise Exception('Not enough entropy left in genome.')
+        self.entropy -= needed
+        quotient, remainder = divmod(self.chromosome, mod)
+        self.log.append({
+            'reason': reason,
+            'mod': mod,
+            'remainder': str(remainder),
+            'before': str(self.chromosome),
+            'after': str(quotient),
```

Comment:
Converting these to strings isn't needed in Python. It was needed in JS because the big int library didn't print out nicely.